aws/rust-runtime/aws-runtime/src/auth/sigv4.rs

```diff
@@ -0,0 +1,430 @@
+    /// Omit the session token from the signature.
+    pub omit_session_token: bool,
+    /// Optional override for the payload to be used in signing.
+    pub payload_override: Option<SignableBody<'static>>,
+    /// Signature type.
+    pub signature_type: HttpSignatureType,
+    /// Whether or not the signature is optional.
+    pub signing_optional: bool,
+    /// Optional expiration (for presigning)
+    pub expires_in: Option<Duration>,
+}
+
+impl Default for SigningOptions {
+    fn default() -> Self {
+        Self {
+            double_uri_encode: true,
+            content_sha256_header: false,
+            normalize_uri_path: true,
+            omit_session_token: false,
+            payload_override: None,
+            signature_type: HttpSignatureType::HttpRequestHeaders,
+            signing_optional: false,
+            expires_in: None,
+        }
+    }
+}
+
+/// SigV4 signing configuration for an operation
+///
+/// Although these fields MAY be customized on a per request basis, they are generally static
+/// for a given operation
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SigV4OperationSigningConfig {
+    /// AWS Region to sign for.
+    pub region: SigningRegion,
+    /// AWS Service to sign for.
+    pub service: SigningService,
+    /// Signing options.
+    pub signing_options: SigningOptions,
+}
+
+/// SigV4 HTTP request signer.
+#[derive(Debug, Default)]
+pub struct SigV4HttpRequestSigner;
+
+impl SigV4HttpRequestSigner {
+    /// Creates a new signer instance.
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn settings(operation_config: &SigV4OperationSigningConfig) -> SigningSettings {
+        let mut settings = SigningSettings::default();
+        settings.percent_encoding_mode = if operation_config.signing_options.double_uri_encode {
+            PercentEncodingMode::Double
+        } else {
+            PercentEncodingMode::Single
+        };
+        settings.payload_checksum_kind = if operation_config.signing_options.content_sha256_header {
+            PayloadChecksumKind::XAmzSha256
+        } else {
+            PayloadChecksumKind::NoHeader
+        };
+        settings.uri_path_normalization_mode =
+            if operation_config.signing_options.normalize_uri_path {
+                UriPathNormalizationMode::Enabled
+            } else {
+                UriPathNormalizationMode::Disabled
+            };
+        settings.session_token_mode = if operation_config.signing_options.omit_session_token {
+            SessionTokenMode::Exclude
+        } else {
+            SessionTokenMode::Include
+        };
+        settings.signature_location = match operation_config.signing_options.signature_type {
+            HttpSignatureType::HttpRequestHeaders => SignatureLocation::Headers,
+            HttpSignatureType::HttpRequestQueryParams => SignatureLocation::QueryParams,
+        };
+        settings.expires_in = operation_config.signing_options.expires_in;
+        settings
+    }
```

Comment:
I felt like writing this in a fancy way for some reason:
```suggestion
    fn settings(operation_config: &SigV4OperationSigningConfig) -> SigningSettings {
        let mut settings = SigningSettings::default();
        let SigningOptions {
            double_uri_encode,
            content_sha256_header,
            normalize_uri_path,
            omit_session_token,
            signature_type,
            expires_in,
            ..
        } = operation_config.signing_options;
        settings.percent_encoding_mode = double_uri_encode
            .then_some(PercentEncodingMode::Double)
            .unwrap_or(PercentEncodingMode::Single);
        settings.payload_checksum_kind = content_sha256_header
            .then_some(PayloadChecksumKind::XAmzSha256)
            .unwrap_or(PayloadChecksumKind::NoHeader);
        settings.uri_path_normalization_mode = normalize_uri_path
            .then_some(UriPathNormalizationMode::Enabled)
            .unwrap_or(UriPathNormalizationMode::Disabled);
        settings.session_token_mode = omit_session_token
            .then_some(SessionTokenMode::Exclude)
            .unwrap_or(SessionTokenMode::Include);
        settings.signature_location = match signature_type {
            HttpSignatureType::HttpRequestHeaders => SignatureLocation::Headers,
            HttpSignatureType::HttpRequestQueryParams => SignatureLocation::QueryParams,
        };
        settings.expires_in = expires_in;
        settings
    }
```

Reply:
I forgot they stabilized `then_some`. Turns out clippy doesn't like it 😆 
```
warning: use of `.then_some(..).unwrap_or(..)` can be written more clearly with `if .. else ..`
   --> aws-runtime/src/auth/sigv4.rs:145:39
    |
145 |           settings.session_token_mode = operation_config
    |  _______________________________________^
146 | |             .signing_options
147 | |             .omit_session_token
148 | |             .then_some(SessionTokenMode::Exclude)
149 | |             .unwrap_or(SessionTokenMode::Include);
    | |_________________________________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#obfuscated_if_else
help: try
    |
145 ~         settings.session_token_mode = if operation_config
146 +             .signing_options
147 ~             .omit_session_token { SessionTokenMode::Exclude } else { SessionTokenMode::Include };
    |
```

Reply:
🤷‍♀️ Clippy knows best, I suppose.